src/frequenz/sdk/timeseries/logical_meter/_formula_engine.py

```diff
@@ -0,0 +1,454 @@
+"""A formula engine that can apply formulas on streaming data.
+
+Copyright
+Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+License
+MIT
+"""
+
+from __future__ import annotations
+
+import asyncio
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum
+from typing import Dict, List, Optional, Set
+
+from frequenz.channels import Receiver, Sender
+
+from ...actor import ChannelRegistry, ComponentMetricRequest
+from ...microgrid.component import ComponentMetricId
+from ...timeseries import Sample
+from ._formula_steps import (
+    Adder,
+    Divider,
+    FormulaStep,
+    MetricFetcher,
+    Multiplier,
+    OpenParen,
+    Subtractor,
+)
+
+_operator_precedence = {
+    "(": 0,
+    "/": 1,
+    "*": 2,
+    "-": 3,
+    "+": 4,
+    ")": 5,
+}
+
+
+class StringIter:
```

Comment:
I don't know,  I like the current approach better,  because we'd need a constructor anyway,  and one of the fields is exposed,  but not part of the constructor.  Otherwise we'd be making code harder to read in order to save a few lines.

Reply:
For me the advantage of having `@dataclass(frozen=True)` is not the saving of a couple of lines but to make it very clear that it is an immutable class (when I read the decorator I immediately know that without reading all the code for it). But I agree that if it is not **really** a `dataclass`, but just an *immutable* class and you need to do some tricks to hide some attributes, then it might not be worth it.

Reply:
Ah, the class was not even really immutable though, because `_pos` changes, so never mind.